src/workerd/io/io-context.h

```diff
@@ -1029,15 +1029,20 @@ jsg::Promise<IoContext::MaybeIoOwn<addIoOwn, T>> IoContext::awaitIoImpl(
                   maybeException = kj::mv(maybeException),
                   maybeAsyncContext = kj::mv(maybeAsyncContext)](Worker::Lock& lock) mutable {
         jsg::AsyncContextFrame::Scope asyncScope(lock, maybeAsyncContext);
-        KJ_IF_SOME(exception, maybeException) {
-          resolver.reject(lock, kj::mv(exception));
-        } else {
-          resolver.resolve(lock);
-        }
+        resolver.resolve(kj::mv(maybeException));
       }, kj::mv(ilOrCs));
     }));
+
+    return jsPromise.then(js, [](jsg::Lock& js, kj::Maybe<kj::Exception>&& maybeEx) -> Result {
+      KJ_IF_SOME(e, maybeEx) {
+        js.throwException(kj::mv(e));
+      }
+    });
   } else {
     typedef kj::OneOf<T, kj::Exception> ResultOrException;
+    typedef kj::OneOf<Result, kj::Exception> IoResultOrException;
```

Comment:
non-blocking nit... naming here (`ResultOrException` and `IoResultOrException`) is a bit unfortunate, but no better suggestions.